{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "bcdc0673",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(dndscv)\n",
    "library(VariantAnnotation)\n",
    "library(GenomicRanges)\n",
    "library(dndscv)\n",
    "library(TxDb.Hsapiens.UCSC.hg38.knownGene)\n",
    "library(GenomicFeatures)\n",
    "library(dplyr)\n",
    "\n",
    "load(\"/uufs/chpc.utah.edu/common/HIPAA/u1264408/u1264408/Git/SEMIColon/data/refcds_GRCh38_hg38.rda\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "8a232b79",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene\n",
    "\n",
    "tx_gr <- transcripts(txdb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65f4f274",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "tx_meta <- as.data.frame(mcols(tx_gr))\n",
    "\n",
    "tx2gene <- data.frame(\n",
    "  tx = tx_meta$tx_name,\n",
    "  gene_id = tx_meta$tx_id,\n",
    "  stringsAsFactors = FALSE\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "22ce776a",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Drop transcripts without gene IDs\n",
    "tx2gene <- tx2gene[!is.na(tx2gene$gene_id), ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "1fc4c759",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "tx_tbl <- data.frame(\n",
    "  tx = names(cds_by_tx),\n",
    "  cds_len = cds_lengths,\n",
    "  stringsAsFactors = FALSE\n",
    ")\n",
    "\n",
    "tx_tbl <- merge(tx_tbl, tx2gene, by = \"tx\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "bf70ee45",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "best_tx <- tx_tbl %>%\n",
    "  group_by(gene_id) %>%\n",
    "  slice_max(cds_len, n = 1) %>%\n",
    "  ungroup()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "911d64ca",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "cds_table <- do.call(\n",
    "  rbind,\n",
    "  lapply(seq_len(nrow(best_tx)), function(i) {\n",
    "\n",
    "    tx <- best_tx$tx[i]\n",
    "    gene_id <- best_tx$gene_id[i]\n",
    "    gr <- cds_by_tx[[tx]]\n",
    "\n",
    "    data.frame(\n",
    "      gene_id   = gene_id,\n",
    "      gene_name = gene_id,   # map to symbols later if desired\n",
    "      cds_id    = tx,\n",
    "      chr       = as.character(seqnames(gr)),\n",
    "      start     = start(gr),\n",
    "      end       = end(gr),\n",
    "      cds_start = start(gr),\n",
    "      cds_end   = end(gr),\n",
    "      length    = sum(width(gr)),\n",
    "      strand    = as.character(strand(gr)),\n",
    "      stringsAsFactors = FALSE\n",
    "    )\n",
    "  })\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "d13c7582",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "vcf_to_dndscv_strict <- function(vcf_file, genome = \"hg38\") {\n",
    "\n",
    "  vcf <- readVcf(vcf_file, genome = genome)\n",
    "\n",
    "  # Expand multiallelics\n",
    "  vcf <- expand(vcf)\n",
    "\n",
    "  rr  <- rowRanges(vcf)\n",
    "  ref <- as.character(ref(vcf))\n",
    "  alt <- as.character(unlist(alt(vcf)))\n",
    "\n",
    "  if (length(ref) == 0) {\n",
    "    return(NULL)\n",
    "  }\n",
    "\n",
    "  df <- data.frame(\n",
    "    chr = as.character(seqnames(rr)),\n",
    "    pos = start(rr),\n",
    "    ref = ref,\n",
    "    mut = alt,\n",
    "    stringsAsFactors = FALSE\n",
    "  )\n",
    "\n",
    "  ## -----------------------------\n",
    "  ## HARD FILTERS\n",
    "  ## -----------------------------\n",
    "\n",
    "  # NA alleles\n",
    "  df <- df[!is.na(df$ref) & !is.na(df$mut), ]\n",
    "\n",
    "  # Uppercase\n",
    "  df$ref <- toupper(df$ref)\n",
    "  df$mut <- toupper(df$mut)\n",
    "\n",
    "  # SNVs only\n",
    "  df <- df[nchar(df$ref) == 1 & nchar(df$mut) == 1, ]\n",
    "\n",
    "  # A/C/G/T only\n",
    "  bases <- c(\"A\", \"C\", \"G\", \"T\")\n",
    "  df <- df[df$ref %in% bases & df$mut %in% bases, ]\n",
    "\n",
    "  # ref != alt\n",
    "  df <- df[df$ref != df$mut, ]\n",
    "\n",
    "  # Chromosome names\n",
    "  df$chr <- sub(\"^chr\", \"\", df$chr)\n",
    "\n",
    "  # ðŸ”‘ CRITICAL: handle empty result\n",
    "  if (nrow(df) == 0) {\n",
    "    return(NULL)\n",
    "  }\n",
    "\n",
    "  # Sample ID\n",
    "  df$sampleID <- sub(\"\\\\.vcf(\\\\.gz)?$\", \"\", basename(vcf_file))\n",
    "\n",
    "  df\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "5bc321b4",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "vcf_dir <- \"/uufs/chpc.utah.edu/common/HIPAA/u1264408/u1264408/Git/SEMIColon/data/output/CellCut/overlap_variants/snvs_all/input\"\n",
    "\n",
    "vcfs <- list.files(vcf_dir, pattern = \"\\\\.vcf(\\\\.gz)?$\", full.names = TRUE)\n",
    "\n",
    "mut_list <- lapply(vcfs, vcf_to_dndscv_strict)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "47b01a95",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(dndscv)\n",
    "library(GenomicRanges)\n",
    "\n",
    "load(\"/uufs/chpc.utah.edu/common/HIPAA/u1264408/u1264408/Git/SEMIColon/data/refcds_GRCh38_hg38.rda\")\n",
    "cds_gr <- do.call(\n",
    "  c,\n",
    "  lapply(RefCDS, function(g) {\n",
    "    if (length(g$cds_start) == 0) return(NULL)\n",
    "\n",
    "    GRanges(\n",
    "      seqnames = rep(g$chr, length(g$cds_start)),\n",
    "      ranges   = IRanges(start = g$cds_start, end = g$cds_end)\n",
    "    )\n",
    "  })\n",
    ")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "d71d50eb",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "mut_gr <- GRanges(\n",
    "  seqnames = mutations$chr,\n",
    "  ranges   = IRanges(start = mutations$pos, end = mutations$pos)\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "dbe02403",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 5 Ã— 5</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>sampleID</th><th scope=col>chr</th><th scope=col>pos</th><th scope=col>ref</th><th scope=col>mut</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>029_DC_T</td><td>1</td><td>3714100</td><td>T</td><td>G</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>029_DC_T</td><td>1</td><td>4475981</td><td>C</td><td>A</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>029_DC_T</td><td>1</td><td>4595543</td><td>G</td><td>T</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>029_DC_T</td><td>1</td><td>5846422</td><td>G</td><td>A</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>029_DC_T</td><td>1</td><td>7908798</td><td>C</td><td>A</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 5 Ã— 5\n",
       "\\begin{tabular}{r|lllll}\n",
       "  & sampleID & chr & pos & ref & mut\\\\\n",
       "  & <chr> & <chr> & <int> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & 029\\_DC\\_T & 1 & 3714100 & T & G\\\\\n",
       "\t2 & 029\\_DC\\_T & 1 & 4475981 & C & A\\\\\n",
       "\t3 & 029\\_DC\\_T & 1 & 4595543 & G & T\\\\\n",
       "\t4 & 029\\_DC\\_T & 1 & 5846422 & G & A\\\\\n",
       "\t5 & 029\\_DC\\_T & 1 & 7908798 & C & A\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 5 Ã— 5\n",
       "\n",
       "| <!--/--> | sampleID &lt;chr&gt; | chr &lt;chr&gt; | pos &lt;int&gt; | ref &lt;chr&gt; | mut &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| 1 | 029_DC_T | 1 | 3714100 | T | G |\n",
       "| 2 | 029_DC_T | 1 | 4475981 | C | A |\n",
       "| 3 | 029_DC_T | 1 | 4595543 | G | T |\n",
       "| 4 | 029_DC_T | 1 | 5846422 | G | A |\n",
       "| 5 | 029_DC_T | 1 | 7908798 | C | A |\n",
       "\n"
      ],
      "text/plain": [
       "  sampleID chr pos     ref mut\n",
       "1 029_DC_T 1   3714100 T   G  \n",
       "2 029_DC_T 1   4475981 C   A  \n",
       "3 029_DC_T 1   4595543 G   T  \n",
       "4 029_DC_T 1   5846422 G   A  \n",
       "5 029_DC_T 1   7908798 C   A  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(mutations, 5)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "8f536fd0",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Drop empty samples\n",
    "mut_list <- Filter(Negate(is.null), mut_list)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "1884ca9d",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "mutations <- do.call(rbind, mut_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "16ca4f2e",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "< table of extent 0 x 0 >"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "NULL"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "stopifnot(\n",
    "  nrow(mutations) > 0,\n",
    "  all(!is.na(mutations$ref)),\n",
    "  all(!is.na(mutations$mut)),\n",
    "  all(nchar(mutations$ref) == 1),\n",
    "  all(nchar(mutations$mut) == 1),\n",
    "  all(mutations$ref %in% c(\"A\",\"C\",\"G\",\"T\")),\n",
    "  all(mutations$mut %in% c(\"A\",\"C\",\"G\",\"T\"))\n",
    ")\n",
    "\n",
    "table(nchar(mutations$ref), nchar(mutations$mut))\n",
    "unique(c(mutations$ref, mutations$mut))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ff843787",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "mut_list <- lapply(vcf_files, vcf_to_dndscv)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0dce418",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "mutations <- do.call(rbind, mut_list)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "77777a82",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "mutations$sampleID <- gsub(\"\\\\.nuclear_all\\\\.intersection$\",\n",
    "                           \"\",\n",
    "                           mutations$sampleID)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4596cfe2",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "mutations$chr <- sub(\"^chr\", \"\", mutations$chr)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9498ed2",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "mutations <- mutations[!is.na(ref) & !is.na(mut)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3a1917ff",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "save.image(file = \"selection.RData\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "de65dc09",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "load(\"selection.RData\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51d9eb7e",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "head(mutations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fc60d1e",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "dndsout <- dndscv(\n",
    "  mutations,\n",
    "  refdb = \"/uufs/chpc.utah.edu/common/HIPAA/u1264408/u1264408/Git/SEMIColon/data/refcds_GRCh38_hg38.rda\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ba2eabed",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "mutations <- mutations[order(mutations$sampleID,\n",
    "                             mutations$chr,\n",
    "                             mutations$pos), ]\n",
    "\n",
    "adjacent <- with(mutations,\n",
    "                 ave(pos, sampleID, chr,\n",
    "                     FUN = function(x) c(FALSE, diff(x) == 1)))\n",
    "\n",
    "mutations_clean <- mutations[!adjacent, ]\n",
    "# there's 1 adjacent mutation recovered\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "682a8a46",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "662248"
      ],
      "text/latex": [
       "662248"
      ],
      "text/markdown": [
       "662248"
      ],
      "text/plain": [
       "[1] 662248"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "   \n",
       "        A     C     G     T\n",
       "  A     0 19380 62874 49354\n",
       "  C 82209     0 21591 98953\n",
       "  G 98089 21461     0 78001\n",
       "  T 49290 61608 19438     0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# After your cleaning step\n",
    "nrow(mutations_clean)\n",
    "table(mutations_clean$ref, mutations_clean$mut)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0350aad2",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "dnds_tmp <- try(\n",
    "  dndscv(\n",
    "    mutations,\n",
    "    refdb = \"/uufs/chpc.utah.edu/common/HIPAA/u1264408/u1264408/Git/SEMIColon/data/refcds_GRCh38_hg38.rda\",\n",
    "    max_muts_per_gene_per_sample = Inf\n",
    "  ),\n",
    "  silent = TRUE\n",
    ")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8d3afb9d",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "mut_per_sample <- table(mutations_clean$sampleID)\n",
    "\n",
    "keep <- names(mut_per_sample[mut_per_sample >= 20])\n",
    "mutations_clean <- mutations_clean[mutations_clean$sampleID %in% keep, ]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f2cce224",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "465"
      ],
      "text/latex": [
       "465"
      ],
      "text/markdown": [
       "465"
      ],
      "text/plain": [
       "[1] 465"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(keep)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "10c2e3b3",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in dndscv(mutations_clean, refdb = \"/uufs/chpc.utah.edu/common/HIPAA/u1264408/u1264408/Git/SEMIColon/data/refcds_GRCh38_hg38.rda\", : unused argument (verbose = TRUE)\n",
     "output_type": "error",
     "traceback": [
      "Error in dndscv(mutations_clean, refdb = \"/uufs/chpc.utah.edu/common/HIPAA/u1264408/u1264408/Git/SEMIColon/data/refcds_GRCh38_hg38.rda\", : unused argument (verbose = TRUE)\nTraceback:\n"
     ]
    }
   ],
   "source": [
    "dndsout <- dndscv(\n",
    "  mutations_clean,\n",
    "  refdb = \"/uufs/chpc.utah.edu/common/HIPAA/u1264408/u1264408/Git/SEMIColon/data/refcds_GRCh38_hg38.rda\", verbose = TRUE\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "344f2f35",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "662249"
      ],
      "text/latex": [
       "662249"
      ],
      "text/markdown": [
       "662249"
      ],
      "text/plain": [
       "[1] 662249"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "465"
      ],
      "text/latex": [
       "465"
      ],
      "text/markdown": [
       "465"
      ],
      "text/plain": [
       "[1] 465"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "nrow(mutations)\n",
    "length(unique(mutations$sampleID))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "50ca9db3",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "is_adjacent\n",
       "     0      1 \n",
       "662248      1 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mutations <- mutations[order(mutations$sampleID,\n",
    "                             mutations$chr,\n",
    "                             mutations$pos), ]\n",
    "\n",
    "is_adjacent <- with(\n",
    "  mutations,\n",
    "  ave(pos, sampleID, chr,\n",
    "      FUN = function(x) {\n",
    "        c(FALSE, diff(x) <= 1)\n",
    "      })\n",
    ")\n",
    "\n",
    "table(is_adjacent)  # <- this will be VERY large\n",
    "\n",
    "mutations_clean <- mutations[!is_adjacent, ]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "bb3826bd",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'T'</li><li>'C'</li><li>'G'</li><li>'A'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'G'\n",
       "\\item 'A'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'T'\n",
       "2. 'C'\n",
       "3. 'G'\n",
       "4. 'A'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"T\" \"C\" \"G\" \"A\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "   \n",
       "         1\n",
       "  1 662249"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mutations <- mutations[!is.na(mutations$ref) & !is.na(mutations$mut), ]\n",
    "mutations$ref <- toupper(mutations$ref)\n",
    "mutations$mut <- toupper(mutations$mut)\n",
    "unique(c(mutations$ref, mutations$mut))\n",
    "table(nchar(mutations$ref), nchar(mutations$mut))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "d3defa47",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'T'</li><li>'C'</li><li>'G'</li><li>'A'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'T'\n",
       "\\item 'C'\n",
       "\\item 'G'\n",
       "\\item 'A'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'T'\n",
       "2. 'C'\n",
       "3. 'G'\n",
       "4. 'A'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"T\" \"C\" \"G\" \"A\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "   \n",
       "         1\n",
       "  1 662249"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "unique(c(mutations$ref, mutations$mut))\n",
    "# \"T\" \"C\" \"G\" \"A\"\n",
    "table(nchar(mutations$ref), nchar(mutations$mut))\n",
    "# 1 x 1 = 662,249\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "c1efed8a",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in dndscv(mutations = mutations, refdb = \"/uufs/chpc.utah.edu/common/HIPAA/u1264408/u1264408/Git/SEMIColon/data/refcds_GRCh38_hg38.rda\", : unused argument (verbose = TRUE)\n",
     "output_type": "error",
     "traceback": [
      "Error in dndscv(mutations = mutations, refdb = \"/uufs/chpc.utah.edu/common/HIPAA/u1264408/u1264408/Git/SEMIColon/data/refcds_GRCh38_hg38.rda\", : unused argument (verbose = TRUE)\nTraceback:\n"
     ]
    }
   ],
   "source": [
    "dndsout <- dndscv(\n",
    "  mutations = mutations,\n",
    "  refdb = \"/uufs/chpc.utah.edu/common/HIPAA/u1264408/u1264408/Git/SEMIColon/data/refcds_GRCh38_hg38.rda\",\n",
    "  verbose = TRUE\n",
    ")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
