{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from cyvcf2 import VCF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf = list(VCF('/Users/quinlan/Documents/Git/SEMIColon/data/PD34199_chr2_chunk2_snps_chunk.vcf'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "variant = vcf[0]\n",
    "# *** variant.POS gives you vcf coordinate; variant.start and variant.end give you bed coordinates)\n",
    "print(variant.CHROM, variant.POS, variant.start, variant.end, variant.REF, variant.ALT)\n",
    "# \"Number of samples with data\"\n",
    "print(variant.INFO.get('NS'))\n",
    "# Total read depth at the locus\"\n",
    "print(variant.INFO.get('DP'))\n",
    "# depth by base\n",
    "print(variant.INFO.get('DPB'))\n",
    "\n",
    "print(variant.format(\"GT\"))\n",
    "\n",
    "# \"Total number of alleles in called genotypes\"\n",
    "print(variant.INFO.get('AN'))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "phases = variant.gt_phases\n",
    "quals = variant.gt_quals,  # numpy array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genotype (GT): ['\\x02\\x02' '\\x02\\x02' '\\x02\\x02' '\\x02\\x02' '\\x02\\x02' '\\x02\\x04'\n",
      " '\\x02\\x02' '\\x02\\x02' '\\x02\\x02' '\\x02\\x02']\n",
      "Genotype Quality (GQ), Phred-scaled marginal probability of the called genotype: None\n",
      "Genotype Likelihood (GL), log10-scaled likelihoods of the data given the called genotype: [[  0.       -5.71957 -60.8677   -5.71957 -60.8677  -60.8677 ]\n",
      " [  0.       -5.11751 -56.8304   -5.11751 -56.8304  -56.8304 ]\n",
      " [  0.       -3.91339 -47.5957   -3.91339 -47.5957  -47.5957 ]\n",
      " [  0.       -4.81648 -52.1476   -2.21863 -49.6177  -49.0157 ]\n",
      " [  0.       -5.41854 -53.0129   -5.41854 -53.0129  -53.0129 ]\n",
      " [ -2.24021   0.      -37.6155   -5.85257 -38.2176  -43.7241 ]\n",
      " [  0.       -7.22472 -82.7304   -7.22472 -82.7304  -82.7304 ]\n",
      " [  0.       -5.11751 -58.0003   -5.11751 -58.0003  -58.0003 ]\n",
      " [  0.       -7.82678 -87.2063   -7.82678 -87.2063  -87.2063 ]\n",
      " [  0.       -8.12781 -91.1764   -8.12781 -91.1764  -91.1764 ]]\n",
      "Read Depth (DP), the total number of reads covering this site: [[19]\n",
      " [17]\n",
      " [13]\n",
      " [19]\n",
      " [18]\n",
      " [14]\n",
      " [24]\n",
      " [17]\n",
      " [26]\n",
      " [27]]\n",
      "Allele Depth (AD), number of observations for each allele (reference and alternate): [[19  0  0]\n",
      " [17  0  0]\n",
      " [13  0  0]\n",
      " [16  0  2]\n",
      " [18  0  0]\n",
      " [12  2  0]\n",
      " [24  0  0]\n",
      " [17  0  0]\n",
      " [26  0  0]\n",
      " [27  0  0]]\n",
      "Reference Allele Observation Count (RO), the number of reads supporting the reference allele: [[19]\n",
      " [17]\n",
      " [13]\n",
      " [16]\n",
      " [18]\n",
      " [12]\n",
      " [24]\n",
      " [17]\n",
      " [26]\n",
      " [27]]\n",
      "Sum of Quality of Reference Observations (QR), sum of quality scores for reference allele reads: [[ 673]\n",
      " [ 628]\n",
      " [ 525]\n",
      " [ 578]\n",
      " [ 586]\n",
      " [ 461]\n",
      " [ 916]\n",
      " [ 641]\n",
      " [ 966]\n",
      " [1010]]\n",
      "Alternate Allele Observation Count (AO), the number of reads supporting the alternate allele: [[0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 2]\n",
      " [0 0]\n",
      " [2 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]\n",
      " [0 0]]\n",
      "Sum of Quality of Alternate Observations (QA), sum of quality scores for alternate allele reads: [[ 0  0]\n",
      " [ 0  0]\n",
      " [ 0  0]\n",
      " [ 0 35]\n",
      " [ 0  0]\n",
      " [68  0]\n",
      " [ 0  0]\n",
      " [ 0  0]\n",
      " [ 0  0]\n",
      " [ 0  0]]\n",
      "Minimum Depth (MIN_DP), minimum read depth across the gVCF block: None\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "# Print the Genotype Quality (GQ)\n",
    "print(f\"Genotype Quality (GQ), Phred-scaled marginal probability of the called genotype: {variant.format('GQ')}\")\n",
    "\n",
    "# Print the Genotype Likelihood (GL)\n",
    "print(f\"Genotype Likelihood (GL), log10-scaled likelihoods of the data given the called genotype: {variant.format('GL')}\")\n",
    "\n",
    "# Print the Read Depth (DP)\n",
    "print(f\"Read Depth (DP), the total number of reads covering this site: {variant.format('DP')}\")\n",
    "\n",
    "# Print the Number of Observations for each Allele (AD)\n",
    "print(f\"Allele Depth (AD), number of observations for each allele (reference and alternate): {variant.format('AD')}\")\n",
    "\n",
    "# Print the Reference Allele Observation Count (RO)\n",
    "print(f\"Reference Allele Observation Count (RO), the number of reads supporting the reference allele: {variant.format('RO')}\")\n",
    "\n",
    "# Print the Sum of Quality of Reference Observations (QR)\n",
    "print(f\"Sum of Quality of Reference Observations (QR), sum of quality scores for reference allele reads: {variant.format('QR')}\")\n",
    "\n",
    "# Print the Alternate Allele Observation Count (AO)\n",
    "print(f\"Alternate Allele Observation Count (AO), the number of reads supporting the alternate allele: {variant.format('AO')}\")\n",
    "\n",
    "# Print the Sum of Quality of Alternate Observations (QA)\n",
    "print(f\"Sum of Quality of Alternate Observations (QA), sum of quality scores for alternate allele reads: {variant.format('QA')}\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# gt_types is array of 0,1,2,3==HOM_REF, HET, UNKNOWN, HOM_ALT\n",
    "genotypes = variant.gt_types\n",
    "alt_depths = variant.gt_alt_depths\n",
    "ref_depths = variant.gt_ref_depths\n",
    "bases = variant.gt_bases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# \"Number of samples with data\"\n",
    "print(\"Number of samples with data:\", variant.INFO.get('NS'))\n",
    "\n",
    "# \"Total read depth at the locus\"\n",
    "print(\"Total read depth at the locus:\", variant.INFO.get('DP'))\n",
    "\n",
    "# \"Total read depth per base\"\n",
    "print(\"Total read depth per base:\", variant.INFO.get('DPB'))\n",
    "\n",
    "# \"Total number of alternate alleles in called genotypes\"\n",
    "print(\"Total number of alternate alleles:\", variant.INFO.get('AC'))\n",
    "\n",
    "# \"Total number of alleles in called genotypes\"\n",
    "print(\"Total number of alleles:\", variant.INFO.get('AN'))\n",
    "\n",
    "# \"Estimated allele frequency\"\n",
    "print(\"Estimated allele frequency:\", variant.INFO.get('AF'))\n",
    "\n",
    "# \"Reference allele observation count (full)\"\n",
    "print(\"Reference allele observation count (full):\", variant.INFO.get('RO'))\n",
    "\n",
    "# \"Alternate allele observation count (full)\"\n",
    "print(\"Alternate allele observation count (full):\", variant.INFO.get('AO'))\n",
    "\n",
    "# \"Reference allele observation count (partial)\"\n",
    "print(\"Reference allele observation count (partial):\", variant.INFO.get('PRO'))\n",
    "\n",
    "# \"Alternate allele observation count (partial)\"\n",
    "print(\"Alternate allele observation count (partial):\", variant.INFO.get('PAO'))\n",
    "\n",
    "# \"Reference allele quality sum\"\n",
    "print(\"Reference allele quality sum (phred):\", variant.INFO.get('QR'))\n",
    "\n",
    "# \"Alternate allele quality sum\"\n",
    "print(\"Alternate allele quality sum (phred):\", variant.INFO.get('QA'))\n",
    "\n",
    "# \"Reference allele quality sum (partial)\"\n",
    "print(\"Reference allele quality sum (partial, phred):\", variant.INFO.get('PQR'))\n",
    "\n",
    "# \"Alternate allele quality sum (partial)\"\n",
    "print(\"Alternate allele quality sum (partial, phred):\", variant.INFO.get('PQA'))\n",
    "\n",
    "# \"Number of reference observations on the forward strand\"\n",
    "print(\"Reference observations (forward strand):\", variant.INFO.get('SRF'))\n",
    "\n",
    "# \"Number of reference observations on the reverse strand\"\n",
    "print(\"Reference observations (reverse strand):\", variant.INFO.get('SRR'))\n",
    "\n",
    "# \"Number of alternate observations on the forward strand\"\n",
    "print(\"Alternate observations (forward strand):\", variant.INFO.get('SAF'))\n",
    "\n",
    "# \"Number of alternate observations on the reverse strand\"\n",
    "print(\"Alternate observations (reverse strand):\", variant.INFO.get('SAR'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# \"Strand balance probability for the reference allele\"\n",
    "print(\"Strand balance probability (reference allele):\", variant.INFO.get('SRP'))\n",
    "\n",
    "# \"Strand balance probability for the alternate allele\"\n",
    "print(\"Strand balance probability (alternate allele):\", variant.INFO.get('SAP'))\n",
    "\n",
    "# \"Allele balance at heterozygous sites\"\n",
    "print(\"Allele balance (heterozygous sites):\", variant.INFO.get('AB'))\n",
    "\n",
    "# \"Allele balance probability at heterozygous sites\"\n",
    "print(\"Allele balance probability (heterozygous sites):\", variant.INFO.get('ABP'))\n",
    "\n",
    "# \"Run length: number of consecutive repeats of the alternate allele in the reference genome\"\n",
    "print(\"Run length (alternate allele):\", variant.INFO.get('RUN'))\n",
    "\n",
    "# \"Read Placement Probability for alternate alleles\"\n",
    "print(\"Read Placement Probability (alternate allele):\", variant.INFO.get('RPP'))\n",
    "\n",
    "# \"Read Placement Probability for reference observations\"\n",
    "print(\"Read Placement Probability (reference observations):\", variant.INFO.get('RPPR'))\n",
    "\n",
    "# \"Reads Placed Left: supporting alternate allele\"\n",
    "print(\"Reads Placed Left (supporting alternate):\", variant.INFO.get('RPL'))\n",
    "\n",
    "# \"Reads Placed Right: supporting alternate allele\"\n",
    "print(\"Reads Placed Right (supporting alternate):\", variant.INFO.get('RPR'))\n",
    "\n",
    "# \"End Placement Probability for alternate alleles\"\n",
    "print(\"End Placement Probability (alternate allele):\", variant.INFO.get('EPP'))\n",
    "\n",
    "# \"End Placement Probability for reference observations\"\n",
    "print(\"End Placement Probability (reference observations):\", variant.INFO.get('EPPR'))\n",
    "\n",
    "# \"Alternate allele depth ratio\"\n",
    "print(\"Alternate allele depth ratio:\", variant.INFO.get('DPRA'))\n",
    "\n",
    "# \"Log odds ratio of best genotype combination\"\n",
    "print(\"Log odds ratio of best genotype combination:\", variant.INFO.get('ODDS'))\n",
    "\n",
    "# \"Number of genotyping iterations to convergence or bailout\"\n",
    "print(\"Genotyping iterations to convergence or bailout:\", variant.INFO.get('GTI'))\n",
    "\n",
    "# \"Type of allele (snp, mnp, ins, del, complex)\"\n",
    "print(\"Type of allele:\", variant.INFO.get('TYPE'))\n",
    "\n",
    "# \"Extended CIGAR representation of alternate allele\"\n",
    "print(\"CIGAR representation of alternate allele:\", variant.INFO.get('CIGAR'))\n",
    "\n",
    "# \"Number of unique non-reference alleles\"\n",
    "print(\"Number of unique non-reference alleles:\", variant.INFO.get('NUMALT'))\n",
    "\n",
    "# \"Mean number of non-reference allele observations per sample\"\n",
    "print(\"Mean number of non-reference allele observations per sample:\", variant.INFO.get('MEANALT'))\n",
    "\n",
    "# \"Allele length\"\n",
    "print(\"Allele length:\", variant.INFO.get('LEN'))\n",
    "\n",
    "# \"Mean mapping quality of observed alternate alleles\"\n",
    "print(\"Mean mapping quality (alternate alleles):\", variant.INFO.get('MQM'))\n",
    "\n",
    "# \"Mean mapping quality of observed reference alleles\"\n",
    "print(\"Mean mapping quality (reference alleles):\", variant.INFO.get('MQMR'))\n",
    "\n",
    "# \"Proportion of observed alternate alleles supported by properly paired reads\"\n",
    "print(\"Proportion of alternate alleles supported by paired reads:\", variant.INFO.get('PAIRED'))\n",
    "\n",
    "# \"Proportion of observed reference alleles supported by properly paired reads\"\n",
    "print(\"Proportion of reference alleles supported by paired reads:\", variant.INFO.get('PAIREDR'))\n",
    "\n",
    "# \"Minimum depth in gVCF output block\"\n",
    "print(\"Minimum depth in gVCF output block:\", variant.INFO.get('MIN_DP'))\n",
    "\n",
    "# \"Last position in gVCF output record\"\n",
    "print(\"Last position in gVCF output record:\", variant.INFO.get('END'))\n",
    "\n",
    "# \"Fraction of alternate allele observations in reads from Illumina\"\n",
    "print(\"Fraction of alternate observations from Illumina reads:\", variant.INFO.get('technology.Illumina'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "# gt_types is array of 0,1,2,3==HOM_REF, HET, UNKNOWN, HOM_ALT\n",
    "HOM_REF=0\n",
    "HET=1\n",
    "UNKNOWN=2\n",
    "HOM_ALT=3\n",
    "\n",
    "# Initialize the dictionary to hold the count of each configuration\n",
    "config_counts = defaultdict(int)\n",
    "\n",
    "# Loop through each variant in the VCF\n",
    "for v in vcf:\n",
    "    # Get the genotypes for the samples\n",
    "    genotypes = v.gt_types\n",
    "\n",
    "    # Assign sample context\n",
    "\n",
    "\n",
    "    if v.QUAL>20 and v.INFO.get(\"NS\") > 1 and v.INFO.get(\"DP\") >= 12:\n",
    "        if blood == youngsperm == oldsperm:\n",
    "            continue\n",
    "        elif blood == UNKNOWN:\n",
    "            continue\n",
    "        elif youngsperm == oldsperm == UNKNOWN:\n",
    "            continue\n",
    "        elif blood == youngsperm and oldsperm == UNKNOWN:\n",
    "            continue\n",
    "        elif blood == oldsperm and youngsperm == UNKNOWN:\n",
    "            continue\n",
    "        else:\n",
    "    # Create a tuple of the genotype configuration\n",
    "            config = (youngsperm, blood, oldsperm)\n",
    "\n",
    "    # Increment the count for this configuration\n",
    "            config_counts[config] += 1\n",
    "\n",
    "\n",
    "# Print the configuration counts\n",
    "for config, count in config_counts.items():\n",
    "    print(f\"{config}: {count} occurrences\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_info_v(v, blood_col, sperm1_col, sperm2_col):\n",
    "    genotypes = v.gt_types\n",
    "    alt_depths = v.gt_alt_depths\n",
    "    ref_depths = v.gt_ref_depths\n",
    "    bases = v.gt_bases\n",
    "\n",
    "    #assign sample context: GT\n",
    "    youngsperm = genotypes[sperm1_col]\n",
    "    blood = genotypes[blood_col]\n",
    "    oldsperm = genotypes[sperm2_col]\n",
    "\n",
    "    #assign sample context: AD\n",
    "    ys_altdepth = alt_depths[sperm1_col]\n",
    "    os_altdepth = alt_depths[sperm2_col]\n",
    "    blood_altdepth = alt_depths[blood_col]\n",
    "\n",
    "    #assign sample context: ref_depths\n",
    "    ys_refdepth = ref_depths[sperm1_col]\n",
    "    os_refdepth = ref_depths[sperm2_col]\n",
    "    blood_refdepth = ref_depths[blood_col]\n",
    "\n",
    "    #assign sample context: ref_depths\n",
    "    ys_bases = bases[sperm1_col]\n",
    "    os_bases = bases[sperm2_col]\n",
    "    blood_bases = bases[blood_col]\n",
    "\n",
    "    return youngsperm, blood, oldsperm, ys_altdepth, os_altdepth, blood_altdepth, ys_refdepth, os_refdepth, blood_refdepth, ys_bases, os_bases, blood_bases\n",
    "\n",
    "def get_interesting(v, youngsperm, blood, oldsperm):\n",
    "    if blood == youngsperm == oldsperm:\n",
    "        return False\n",
    "    elif blood == UNKNOWN:\n",
    "        return False\n",
    "    elif youngsperm == oldsperm == UNKNOWN:\n",
    "        return False\n",
    "    elif blood == youngsperm and oldsperm == UNKNOWN:\n",
    "        return False\n",
    "    elif blood == oldsperm and youngsperm == UNKNOWN:\n",
    "        return False\n",
    "    else:\n",
    "        return True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# gt_types is array of 0,1,2,3==HOM_REF, HET, UNKNOWN, HOM_ALT\n",
    "HOM_REF=0\n",
    "HET=1\n",
    "UNKNOWN=2\n",
    "HOM_ALT=3\n",
    "\n",
    "blood_col = 1\n",
    "sperm1_col = 0\n",
    "sperm2_col = 2\n",
    "\n",
    "# variant.gt_types, variant.gt_ref_depths, variant.gt_alt_depths # numpy arrays\n",
    "# variant.gt_phases, variant.gt_quals, variant.gt_bases # numpy array\n",
    "\n",
    "for v in vcf:\n",
    "    youngsperm, blood, oldsperm, ys_altdepth, os_altdepth, blood_altdepth, ys_refdepth, os_refdepth, blood_refdepth, ys_bases, os_bases, blood_bases = get_info_v(v,\n",
    "                                                                                                                        blood_col, sperm1_col, sperm2_col)\n",
    "    if v.QUAL>20 and v.INFO.get(\"NS\") > 1 and v.INFO.get(\"DP\") > 10 and get_interesting(v, youngsperm, blood, oldsperm):\n",
    "        print(\"blood\", blood_bases, \"alt depth\", blood_altdepth, \"ref depth\", blood_refdepth, \"young sperm\", ys_bases, \"alt depth\", ys_altdepth,\n",
    "                  \"ref depth\", ys_refdepth, \"old sperm\",os_bases, \"alt depth\", os_altdepth, \"ref depth\", os_refdepth)\n",
    "        #print(\"blood\", \"sperm1\", \"sperm2\")\n",
    "        #print(v.INFO.get(\"AC\"))\n",
    "            #print(v.gt_types, ys_altdepth, os_altdepth, v.QUAL)\n",
    "            #print(os_refdepth, ys_refdepth, blood_refdepth)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for v in vcf:\n",
    "    # Get the necessary info\n",
    "    genotypes = v.gt_types\n",
    "\n",
    "    # Assign sample context: GT\n",
    "    youngsperm = genotypes[0]\n",
    "    blood = genotypes[1]\n",
    "    oldsperm = genotypes[2]\n",
    "\n",
    "    # Access FORMAT fields\n",
    "    ro_blood = v.format('RO')[1]\n",
    "    ao_blood = v.format('AO')[1]\n",
    "\n",
    "    ro_youngsperm = v.format('RO')[0]\n",
    "    ao_youngsperm = v.format('AO')[0]\n",
    "\n",
    "    ro_oldsperm = v.format('RO')[2]\n",
    "    ao_oldsperm = v.format('AO')[2]\n",
    "\n",
    "    if v.QUAL>20 and v.INFO.get(\"NS\") > 1 and v.INFO.get(\"DP\") > 10 and get_interesting(v, youngsperm, blood, oldsperm):\n",
    "\n",
    "    # Print the FORMAT fields\n",
    "        print(f\"Blood - RO: {ro_blood}, AO: {ao_blood}\")\n",
    "        print(f\"Young Sperm - RO: {ro_youngsperm}, AO: {ao_youngsperm}\")\n",
    "        print(f\"Old Sperm - RO: {ro_oldsperm}, AO: {ao_oldsperm}\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "filterpy",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
