<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8" />
    <meta name="author" content="Brent Pedersen" />
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/selectize.js/0.12.6/css/selectize.bootstrap3.min.css">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.3.1/css/bootstrap.min.css">
    <script src="https://cdn.plot.ly/plotly-latest.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/selectize.js/0.12.6/js/standalone/selectize.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.3.1/js/bootstrap.bundle.min.js"></script>
    <style>
    html,body { height: 100%; margin: 0px; padding: 0px; }
    .form-control.selectize-control {
        padding: 5px 6px 0px;
        height: unset !important;
    }
    .remove-single {
        color: gray !important;
        top: -1px !important;
        font-size: 20px !important;
    }
    </style>
</head>
<body>

<nav class="navbar navbar-expand-lg navbar-dark p-0" style="background-color:#4b636e !important">
  <a class="navbar-brand ml-2" href="https://github.com/brentp/somalier">somalier</a>
  <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#toggler" aria-controls="toggler" aria-expanded="false" aria-label="Toggle navigation">
    <span class="navbar-toggler-icon"></span>
  </button>

  <div class="collapse navbar-collapse" id="toggler">
    <div class="navbar-nav mr-auto">
    </div>
    <div>
        <div class="input-group" style="width:300px">
            <input type="text" class="form-control" id="sample-search" aria-label="sample-search">
        </div>
    </div>
  </div>
</nav>
<div class="container-fluid h-100">
    <div class="row h-75">
        <div class="col-6 h-100 bg-light pt-2 border-right">
            <div class="row">
                <div class="col-6">
                    <h5>Sample to Sample Relatedness</h5>
                </div>
                <div class="col-6">
                    <input type="checkbox" name="n_scale" id="n_scale" >
                    <label for="n_scale" title="scale the x and y-axis values
                        (except relatedness and concordance) by the number of sites where both samples have a known genotype">Scale by shared-sites</label>
                </div>
            </div>
            <div class="row pb-2">
                <div class="col-6">
                    <div class="input-group">
                        <div class="input-group-prepend">
                            <div class="input-group-text">X Axis</div>
                        </div>
                        <select class="custom-select" id="plotax_select">
                            <option value="shared_hets" title="the number of sites where both samples are heterozygotes">Shared hets</option>
                            <option value="shared_hom_alts" title="the number of sites where both samples are homozygous alternate">Shared hom-alts</option>
                            <option value="concordance" title="a relatedness estimate that is not affected by loss-of-heterozygosity">Homozygous concordance</option>
                            <option value="relatedness" title="relatedness of a pair of samples (2*kinship coefficient")>Relatedness</option>
                            <option value="ibs0" title="the number of sites where one sample is hom-ref and another is hom-alt" selected>IBS0</option>
                            <option value="ibs2" title="the number of sites where the samples have the same genotype">IBS2</option>
                        </select>
                    </div>
                </div>
                <div class="col-6">
                    <div class="input-group">
                        <div class="input-group-prepend">
                            <div class="input-group-text">Y Axis</div>
                        </div>
                        <select class="custom-select" id="plotay_select">
                            <option value="shared_hets" title="the number of sites where both samples are heterozygotes">Shared hets</option>
                            <option value="shared_hom_alts" title="the number of sites where both samples are homozygous alternate">Shared hom-alts</option>
                            <option value="concordance" title="a relatedness estimate that is not affected by loss-of-heterozygosity">Homozygous concordance</option>
                            <option value="relatedness" title="relatedness of a pair of samples (2*kinship coefficient")>Relatedness</option>
                            <option value="ibs0" title="the number of sites where one sample is hom-ref and another is hom-alt">IBS0</option>
                            <option value="ibs2" title="the number of sites where the samples have the same genotype" selected>IBS2</option>
                        </select>
                    </div>
                </div>
            </div>
            <div class="row h-100">
                <div class="col h-100">
                    <div id="plota" style="height: 90%"></div>
                </div>
            </div>
        </div>

        <div class="col-6 h-100 bg-light pt-2">
            <div class="row pt-0 pb-0">
                <div class="col pt-0 pb-0">
                    <h5>Sample Depth Metrics</h5>
                </div>
            </div>
            <div class="row pb-2">
                <div class="col-6">
                    <div class="input-group">
                        <div class="input-group-prepend">
                            <div class="input-group-text">X Axis</div>
                        </div>
                        <select class="custom-select" id="plotbx_select">
                            <option value="gt_depth_mean">Mean depth of genotyped sites</option>
                            <option value="depth_mean">Mean depth of all sites</option>
                            <option value="ab_mean">Mean allele balance</option>
                            <option value="pct_other_alleles">% reads with neither REF nor ALT</option>
                            <option value="heterozygosity">Heterozygosity rate</option>
                            <option value="n_hom_ref">Number of 0/0 sites</option>
                            <option value="n_het">Number of 0/1 sites</option>
                            <option value="n_hom_alt">Number of 1/1 sites</option>
                            <option value="n_unknown">Number of unknown sites</option>
                            <option value="n_known">Number of known sites</option>
                            <option value="p_middling_ab">Proportion sites with AB &lt; 0.1 or AB &gt; 0.9</option>
                            <option value="x_depth_mean">Scaled mean depth on X</option>
                            <option value="x_hom_ref">Number of 0/0 sites on X</option>
                            <option value="x_het" selected >Number of 0/1 sites on X</option>
                            <option value="x_hom_alt">Number of 1/1 sites on X</option>
                            <option value="y_depth_mean">Scaled mean depth on Y</option>
                        </select>
                    </div>
                </div>
                <div class="col-6">
                    <div class="input-group">
                        <div class="input-group-prepend">
                            <div class="input-group-text">Y Axis</div>
                        </div>
                        <select class="custom-select" id="plotby_select">
                            <option value="gt_depth_mean">Mean depth of genotyped sites</option>
                            <option value="depth_mean">Mean depth of all sites</option>
                            <option value="ab_mean">Mean allele balance</option>
                            <option value="pct_other_alleles">% reads with neither REF nor ALT</option>
                            <option value="heterozygosity">Heterozygosity rate</option>
                            <option value="n_hom_ref">Number of 0/0 sites</option>
                            <option value="n_het">Number of 0/1 sites</option>
                            <option value="n_hom_alt">Number of 1/1 sites</option>
                            <option value="n_unknown">Number of unknown sites</option>
                            <option value="n_known">Number of known sites</option>
                            <option value="p_middling_ab">Proportion sites with AB &lt; 0.1 or AB &gt; 0.9</option>
                            <option value="x_depth_mean" selected>Scaled mean depth on X</option>
                            <option value="x_hom_ref">Number of 0/0 sites on X</option>
                            <option value="x_het">Number of 0/1 sites on X</option>
                            <option value="x_hom_alt">Number of 1/1 sites on X</option>
                            <option value="y_depth_mean">Scaled mean depth on Y</option>
                        </select>
                    </div>
                </div>
            </div>
            <div class="row h-100">
                <div class="col h-100">
                    <div id="plotb" style="height: 90%"></div>
                </div>
            </div>
        </div>
    </div>
    <div class="row h-25">
        <div class="col-12 h-100 bg-light pt-2 border-right">
            Choose pre-set selection for sample plot:
            <ul>
            <li><a href="javascript:set_plotb('x_het', 'x_depth_mean')">Sex QC: Number of 0/1 sites on X vs scaled mean depth on chrX</a></li>

            <li><a href="javascript:set_plotb('x_depth_mean', 'y_depth_mean')">Sex ploidy QC: Scaled mean depth on chrX vs. chrY</a></li>
            <li><a href="javascript:set_plotb('n_unknown', 'p_middling_ab')">General QC: number of sites with unknown genotype vs. proportion of hets with allele balance outside of 0.1 - 0.9</a></li>
            <li><a href="javascript:set_plotb('gt_depth_mean', 'ab_mean')">Depth and allele QC: mean depth vs. mean het allele-balance</a></li>
            </ul>
        </div>
    </div>
</div>

<script>
var nan = NaN; // hack to support json dumped with NaN values.
var inf = Infinity
var sample_data = [{"sample":"D091_IX8Y012_19841X13","sex":"unknown","gt_depth_mean":7.0,"depth_mean":0.9112402208927768,"ab_mean":-0.9884393063583816,"pct_other_alleles":0.2335528439537733,"n_hom_ref":0,"n_het":0,"n_hom_alt":1,"n_unknown":17383,"n_known":1,"p_middling_ab":0.0,"x_depth_mean":0.0,"x_hom_ref":0,"x_het":0,"x_hom_alt":0,"y_depth_mean":0.0},
{"sample":"D091_IX8Y012_19841X3","sex":"unknown","gt_depth_mean":0.0,"depth_mean":0.2244592728946158,"ab_mean":-1.0,"pct_other_alleles":0.3433476394849778,"n_hom_ref":0,"n_het":0,"n_hom_alt":0,"n_unknown":17384,"n_known":0,"p_middling_ab":0.0,"x_depth_mean":"nan","x_hom_ref":0,"x_het":0,"x_hom_alt":0,"y_depth_mean":"nan"},
{"sample":"D091_IX8Y012_19841X8","sex":"unknown","gt_depth_mean":7.153846153846154,"depth_mean":1.388288080994011,"ab_mean":-0.9713345864661658,"pct_other_alleles":0.1625828916711476,"n_hom_ref":5,"n_het":5,"n_hom_alt":1,"n_unknown":17373,"n_known":11,"p_middling_ab":0.0,"x_depth_mean":0.0,"x_hom_ref":0,"x_het":0,"x_hom_alt":0,"y_depth_mean":0.0},
{"sample":"D100_VG6U010_19841X14","sex":"unknown","gt_depth_mean":7.560000000000001,"depth_mean":1.605614358030377,"ab_mean":-0.9623875061941534,"pct_other_alleles":0.2458367078934967,"n_hom_ref":8,"n_het":6,"n_hom_alt":6,"n_unknown":17364,"n_known":20,"p_middling_ab":0.0,"x_depth_mean":0.0,"x_hom_ref":0,"x_het":0,"x_hom_alt":0,"y_depth_mean":0.0},
{"sample":"D100_VG6U010_19841X4","sex":"unknown","gt_depth_mean":0.0,"depth_mean":0.09917165209387885,"ab_mean":0.0,"pct_other_alleles":0.4270896888346561,"n_hom_ref":0,"n_het":0,"n_hom_alt":0,"n_unknown":17384,"n_known":0,"p_middling_ab":0.0,"x_depth_mean":"nan","x_hom_ref":0,"x_het":0,"x_hom_alt":0,"y_depth_mean":"nan"},
{"sample":"D100_VG6U010_19841X9","sex":"unknown","gt_depth_mean":9.797160731472559,"depth_mean":8.433214450069046,"ab_mean":0.2016775181681092,"pct_other_alleles":0.2056253157068847,"n_hom_ref":3378,"n_het":3913,"n_hom_alt":4021,"n_unknown":6072,"n_known":11312,"p_middling_ab":0.002009300189448304,"x_depth_mean":1.676870664762702,"x_hom_ref":20,"x_het":0,"x_hom_alt":36,"y_depth_mean":1.480020630203602},
{"sample":"D127_V5ET001_19841X12","sex":"unknown","gt_depth_mean":9.199999999999999,"depth_mean":0.02973999079613445,"ab_mean":-0.3999999999999999,"pct_other_alleles":0.0,"n_hom_ref":1,"n_het":0,"n_hom_alt":4,"n_unknown":17379,"n_known":5,"p_middling_ab":0.0,"x_depth_mean":0.0,"x_hom_ref":0,"x_het":0,"x_hom_alt":0,"y_depth_mean":0.0},
{"sample":"D127_V5ET001_19841X2","sex":"unknown","gt_depth_mean":0.0,"depth_mean":0.04078462954440876,"ab_mean":0.0,"pct_other_alleles":0.1432664756446992,"n_hom_ref":0,"n_het":0,"n_hom_alt":0,"n_unknown":17384,"n_known":0,"p_middling_ab":0.0,"x_depth_mean":"nan","x_hom_ref":0,"x_het":0,"x_hom_alt":0,"y_depth_mean":"nan"},
{"sample":"D127_V5ET001_19841X7","sex":"unknown","gt_depth_mean":0.0,"depth_mean":0.1290266912103095,"ab_mean":0.0,"pct_other_alleles":0.1417769376181473,"n_hom_ref":0,"n_het":0,"n_hom_alt":0,"n_unknown":17384,"n_known":0,"p_middling_ab":0.0,"x_depth_mean":"nan","x_hom_ref":0,"x_het":0,"x_hom_alt":0,"y_depth_mean":"nan"},
{"sample":"D130_WN11-001_19841X11","sex":"unknown","gt_depth_mean":7.0,"depth_mean":1.061378278877135,"ab_mean":-0.9801324503311258,"pct_other_alleles":0.2258938563939449,"n_hom_ref":2,"n_het":0,"n_hom_alt":2,"n_unknown":17380,"n_known":4,"p_middling_ab":0.0,"x_depth_mean":0.0,"x_hom_ref":0,"x_het":0,"x_hom_alt":0,"y_depth_mean":0.0},
{"sample":"D130_WN11-001_19841X1","sex":"unknown","gt_depth_mean":0.0,"depth_mean":0.1394385641969631,"ab_mean":0.0,"pct_other_alleles":0.3097345132743367,"n_hom_ref":0,"n_het":0,"n_hom_alt":0,"n_unknown":17384,"n_known":0,"p_middling_ab":0.0,"x_depth_mean":"nan","x_hom_ref":0,"x_het":0,"x_hom_alt":0,"y_depth_mean":"nan"},
{"sample":"D130_WN11-001_19841X6","sex":"unknown","gt_depth_mean":9.646629454701626,"depth_mean":8.055855959503031,"ab_mean":0.1543345092635259,"pct_other_alleles":0.1833858457644961,"n_hom_ref":3142,"n_het":3474,"n_hom_alt":3859,"n_unknown":6909,"n_known":10475,"p_middling_ab":0.002181150269773849,"x_depth_mean":1.669522316942035,"x_hom_ref":16,"x_het":0,"x_hom_alt":22,"y_depth_mean":1.451284105577063},
{"sample":"D212_92N8014_19841X10","sex":"unknown","gt_depth_mean":7.272727272727272,"depth_mean":1.254832029452368,"ab_mean":-0.9653623403623403,"pct_other_alleles":0.1995911340649884,"n_hom_ref":2,"n_het":5,"n_hom_alt":3,"n_unknown":17374,"n_known":10,"p_middling_ab":0.0,"x_depth_mean":0.0,"x_hom_ref":0,"x_het":0,"x_hom_alt":0,"y_depth_mean":0.0},
{"sample":"D212_92N8014_19841X15","sex":"unknown","gt_depth_mean":7.791342952275254,"depth_mean":3.235618959963179,"ab_mean":-0.7379231655763879,"pct_other_alleles":0.1690220904910515,"n_hom_ref":228,"n_het":224,"n_hom_alt":316,"n_unknown":16616,"n_known":768,"p_middling_ab":5.752085130859936e-05,"x_depth_mean":2.053561253561252,"x_hom_ref":0,"x_het":0,"x_hom_alt":1,"y_depth_mean":0.0},
{"sample":"D212_92N8014_19841X5","sex":"unknown","gt_depth_mean":0.0,"depth_mean":0.1648642429820523,"ab_mean":0.0,"pct_other_alleles":0.2271006813020441,"n_hom_ref":0,"n_het":0,"n_hom_alt":0,"n_unknown":17384,"n_known":0,"p_middling_ab":0.0,"x_depth_mean":"nan","x_hom_ref":0,"x_het":0,"x_hom_alt":0,"y_depth_mean":"nan"}]
var input = [{"expected_relatedness":0.0,"text":["D091_IX8Y012_19841X13<br>D091_IX8Y012_19841X3","D091_IX8Y012_19841X13<br>D091_IX8Y012_19841X8","D091_IX8Y012_19841X13<br>D100_VG6U010_19841X14","D091_IX8Y012_19841X13<br>D100_VG6U010_19841X4","D091_IX8Y012_19841X13<br>D100_VG6U010_19841X9","D091_IX8Y012_19841X13<br>D127_V5ET001_19841X12","D091_IX8Y012_19841X13<br>D127_V5ET001_19841X2","D091_IX8Y012_19841X13<br>D127_V5ET001_19841X7","D091_IX8Y012_19841X13<br>D130_WN11-001_19841X11","D091_IX8Y012_19841X13<br>D130_WN11-001_19841X1","D091_IX8Y012_19841X13<br>D130_WN11-001_19841X6","D091_IX8Y012_19841X13<br>D212_92N8014_19841X10","D091_IX8Y012_19841X13<br>D212_92N8014_19841X15","D091_IX8Y012_19841X13<br>D212_92N8014_19841X5","D091_IX8Y012_19841X3<br>D091_IX8Y012_19841X8","D091_IX8Y012_19841X3<br>D100_VG6U010_19841X14","D091_IX8Y012_19841X3<br>D100_VG6U010_19841X4","D091_IX8Y012_19841X3<br>D100_VG6U010_19841X9","D091_IX8Y012_19841X3<br>D127_V5ET001_19841X12","D091_IX8Y012_19841X3<br>D127_V5ET001_19841X2","D091_IX8Y012_19841X3<br>D127_V5ET001_19841X7","D091_IX8Y012_19841X3<br>D130_WN11-001_19841X11","D091_IX8Y012_19841X3<br>D130_WN11-001_19841X1","D091_IX8Y012_19841X3<br>D130_WN11-001_19841X6","D091_IX8Y012_19841X3<br>D212_92N8014_19841X10","D091_IX8Y012_19841X3<br>D212_92N8014_19841X15","D091_IX8Y012_19841X3<br>D212_92N8014_19841X5","D091_IX8Y012_19841X8<br>D100_VG6U010_19841X14","D091_IX8Y012_19841X8<br>D100_VG6U010_19841X4","D091_IX8Y012_19841X8<br>D100_VG6U010_19841X9","D091_IX8Y012_19841X8<br>D127_V5ET001_19841X12","D091_IX8Y012_19841X8<br>D127_V5ET001_19841X2","D091_IX8Y012_19841X8<br>D127_V5ET001_19841X7","D091_IX8Y012_19841X8<br>D130_WN11-001_19841X11","D091_IX8Y012_19841X8<br>D130_WN11-001_19841X1","D091_IX8Y012_19841X8<br>D130_WN11-001_19841X6","D091_IX8Y012_19841X8<br>D212_92N8014_19841X10","D091_IX8Y012_19841X8<br>D212_92N8014_19841X15","D091_IX8Y012_19841X8<br>D212_92N8014_19841X5","D100_VG6U010_19841X14<br>D100_VG6U010_19841X4","D100_VG6U010_19841X14<br>D100_VG6U010_19841X9","D100_VG6U010_19841X14<br>D127_V5ET001_19841X12","D100_VG6U010_19841X14<br>D127_V5ET001_19841X2","D100_VG6U010_19841X14<br>D127_V5ET001_19841X7","D100_VG6U010_19841X14<br>D130_WN11-001_19841X11","D100_VG6U010_19841X14<br>D130_WN11-001_19841X1","D100_VG6U010_19841X14<br>D130_WN11-001_19841X6","D100_VG6U010_19841X14<br>D212_92N8014_19841X10","D100_VG6U010_19841X14<br>D212_92N8014_19841X15","D100_VG6U010_19841X14<br>D212_92N8014_19841X5","D100_VG6U010_19841X4<br>D100_VG6U010_19841X9","D100_VG6U010_19841X4<br>D127_V5ET001_19841X12","D100_VG6U010_19841X4<br>D127_V5ET001_19841X2","D100_VG6U010_19841X4<br>D127_V5ET001_19841X7","D100_VG6U010_19841X4<br>D130_WN11-001_19841X11","D100_VG6U010_19841X4<br>D130_WN11-001_19841X1","D100_VG6U010_19841X4<br>D130_WN11-001_19841X6","D100_VG6U010_19841X4<br>D212_92N8014_19841X10","D100_VG6U010_19841X4<br>D212_92N8014_19841X15","D100_VG6U010_19841X4<br>D212_92N8014_19841X5","D100_VG6U010_19841X9<br>D127_V5ET001_19841X12","D100_VG6U010_19841X9<br>D127_V5ET001_19841X2","D100_VG6U010_19841X9<br>D127_V5ET001_19841X7","D100_VG6U010_19841X9<br>D130_WN11-001_19841X11","D100_VG6U010_19841X9<br>D130_WN11-001_19841X1","D100_VG6U010_19841X9<br>D130_WN11-001_19841X6","D100_VG6U010_19841X9<br>D212_92N8014_19841X10","D100_VG6U010_19841X9<br>D212_92N8014_19841X15","D100_VG6U010_19841X9<br>D212_92N8014_19841X5","D127_V5ET001_19841X12<br>D127_V5ET001_19841X2","D127_V5ET001_19841X12<br>D127_V5ET001_19841X7","D127_V5ET001_19841X12<br>D130_WN11-001_19841X11","D127_V5ET001_19841X12<br>D130_WN11-001_19841X1","D127_V5ET001_19841X12<br>D130_WN11-001_19841X6","D127_V5ET001_19841X12<br>D212_92N8014_19841X10","D127_V5ET001_19841X12<br>D212_92N8014_19841X15","D127_V5ET001_19841X12<br>D212_92N8014_19841X5","D127_V5ET001_19841X2<br>D127_V5ET001_19841X7","D127_V5ET001_19841X2<br>D130_WN11-001_19841X11","D127_V5ET001_19841X2<br>D130_WN11-001_19841X1","D127_V5ET001_19841X2<br>D130_WN11-001_19841X6","D127_V5ET001_19841X2<br>D212_92N8014_19841X10","D127_V5ET001_19841X2<br>D212_92N8014_19841X15","D127_V5ET001_19841X2<br>D212_92N8014_19841X5","D127_V5ET001_19841X7<br>D130_WN11-001_19841X11","D127_V5ET001_19841X7<br>D130_WN11-001_19841X1","D127_V5ET001_19841X7<br>D130_WN11-001_19841X6","D127_V5ET001_19841X7<br>D212_92N8014_19841X10","D127_V5ET001_19841X7<br>D212_92N8014_19841X15","D127_V5ET001_19841X7<br>D212_92N8014_19841X5","D130_WN11-001_19841X11<br>D130_WN11-001_19841X1","D130_WN11-001_19841X11<br>D130_WN11-001_19841X6","D130_WN11-001_19841X11<br>D212_92N8014_19841X10","D130_WN11-001_19841X11<br>D212_92N8014_19841X15","D130_WN11-001_19841X11<br>D212_92N8014_19841X5","D130_WN11-001_19841X1<br>D130_WN11-001_19841X6","D130_WN11-001_19841X1<br>D212_92N8014_19841X10","D130_WN11-001_19841X1<br>D212_92N8014_19841X15","D130_WN11-001_19841X1<br>D212_92N8014_19841X5","D130_WN11-001_19841X6<br>D212_92N8014_19841X10","D130_WN11-001_19841X6<br>D212_92N8014_19841X15","D130_WN11-001_19841X6<br>D212_92N8014_19841X5","D212_92N8014_19841X10<br>D212_92N8014_19841X15","D212_92N8014_19841X10<br>D212_92N8014_19841X5","D212_92N8014_19841X15<br>D212_92N8014_19841X5"],"ibs0":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,3,0,1,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,738,1,50,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,53,0,0,0,0],"ibs2":[0,0,0,0,1,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,5,0,0,0,0,0,3,0,0,0,0,15,0,0,0,0,0,6,1,1,0,0,0,0,0,0,0,0,0,0,0,4,0,0,2,0,3242,6,255,0,0,0,0,0,4,0,3,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,3,0,0,0,0,0,0,0,1,233,0,0,0,0],"shared_hets":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,3,0,0,0,0,0,1,0,0,0,0,3,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,874,2,68,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,59,0,0,0,0],"shared_hom_alts":[0,0,0,0,1,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,5,0,0,0,0,0,4,0,0,0,0,0,0,0,0,0,0,0,0,0,3,0,0,1,0,1372,2,110,0,0,0,0,0,3,0,2,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,2,0,0,0,0,0,0,0,1,109,0,0,0,0],"concordance":[0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,-2.0,0.0,0.0,0.0,0.0,0.0,-2.0,0.0,0.0,0.0,0.0,0.8333333134651184,0.0,0.0,0.0,0.0,0.0,-0.3333333432674408,0.0,-0.3333333432674408,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.25,0.0,0.0,0.5,0.0,-0.02694998681545258,0.0,0.03164556995034218,0.0,0.0,0.0,0.0,0.0,0.75,0.0,0.5,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.3333333432674408,0.009493670426309109,0.0,0.0,0.0,0.0],"relatedness":["nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","nan",0.25,"nan","nan","nan","nan","nan",-0.3333333432674408,"nan",0.0,"nan","nan",1.0,"nan","nan","nan","nan","nan",-2.400000095367432,1.0,-4.0,"nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","-inf","nan","nan",0.0,"nan",-0.2573749423027039,0.0,-0.1849710941314697,"nan","nan","nan","nan","nan","nan","nan",0.0,"nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","nan","nan",0.0,-0.291925460100174,"nan","nan","nan","nan"],"n":[0,0,0,0,1,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,8,0,0,0,0,0,8,0,1,0,0,15,0,0,0,0,0,14,1,3,0,0,0,0,0,0,0,0,0,0,0,5,0,0,3,0,6910,8,515,0,0,0,0,0,4,0,4,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,3,0,0,0,0,0,0,0,7,490,0,0,0,0]}]

// ibs0 and ibs2 are counts. on page-load, we change them to scale by n
    /*
for(var i=0; i< input.length; i++){
    var rel = input[i];
    for(var j=0; j<rel.ibs0.length; j++){
        rel.ibs0[j] /= rel.n[j];
        rel.ibs2[j] /= rel.n[j];
    }
}
*/


var colors = ['rgba(55,126,184,0.7)', 'rgba(228,26,28,0.7)', 'rgba(152,78,163,0.7)', 'rgba(255,127,0,0.7)', 'rgba(166,86,40,0.7)', 'rgba(247,129,191,0.7)', 'rgba(77,175,74,0.7)',]
var size
if (sample_data.length > 700) {
    size = 8
} else if (sample_data.length > 200) {
    size = 9
} else if (sample_data.length > 50) {
    size = 10
} else if (sample_data.length > 20) {
    size = 12
} else {
    size = 15
}

function set_plotb(x, y) {
    jQuery('#plotbx_select').val(x)
    jQuery('#plotby_select').val(y)
    jQuery('#plotby_select, #plotbx_select').trigger('change')
}

function set_xy_data_by_group(input, metric, is_x, scale) {
    let l = {"0.00": "unrelated", "0.50": "parent-child", "0.49": "sib-sib", "1.00": "identical"}
    for(var i = 0; i < input.length; i++) {
        var c = input[i]
        if(c.expected_relatedness != undefined) {
                c.name = l[c.expected_relatedness.toFixed(2)] || c.expected_relatedness.toFixed(3)
        }
        if(is_x) {
            c.x = c[metric]
            c.showlegend = i < input.length - 1 && c.x.length > 0;
            if(scale && c.x && metric != "relatedness" && metric != "concordance") {
                c.x = c.x.slice()
                for(var k=0; k< c.x.length; k++){
                   c.x[k] = c.x[k] / c.n[k]
               }
            }
        } else {
           c.y = c[metric]
           c.showlegend =i < input.length - 1 &&  c.y.length > 0;
            if(scale && c.y && metric != "concordance" && metric != "relatedness") {
                c.y = c.y.slice()
                for(var k=0; k< c.y.length; k++){
                   c.y[k] = c.y[k] / c.n[k]
               }
            }
        }

        c.type = 'scattergl'
        c.mode = 'markers'
        if(i < input.length - 1 ){
           c.marker = {size: size, color: colors[i]}
        }

    }
}


var layout_a = {
    autosize: true,
    margin: {t: 30, pad: 0},
    xaxis: {
        title: jQuery("#plotax_select option:selected").text(),
    },
    yaxis: {
        title: jQuery("#plotay_select option:selected").text(),
    },
    hovermode: 'closest',
    showlegend: true,
    legend: {
        xanchor: "right",
        yanchor: "top",
        y: 1,
        x: 1,
        orientation: "h",
        borderwidth: 1,
        bordercolor: '#eeeeee'
    },
}

var layout_b = {
    autosize: true,
    margin: {t: 30, pad: 0},
    xaxis: {
        title: jQuery("#plotbx_select option:selected").text(),
    },
    yaxis: {
        title: jQuery("#plotby_select option:selected").text(),
    },
    hovermode: 'closest',
    showlegend: true,
    legend: {
        xanchor: "right",
        yanchor: "top",
        y: 1,
        x: 1,
        orientation: "h",
        borderwidth: 1,
        bordercolor: '#eeeeee'
    },
}

input.push({mode:"markers", type:'scattergl', hoverinfo:'text+x+y', showlegend:false, name:'hovered', x:[], y:[], hovertext:[], marker:{size:size+3, opacity: 1, color:'#000000'}})
set_xy_data_by_group(input, jQuery('#plotax_select').val(), true, false)
set_xy_data_by_group(input, jQuery('#plotay_select').val(), false, false)

// append final trace to hold plot B hover action markers

var traces_b = [
    // gray
    {x:[], y:[], text: [], type: 'scatter', mode: 'markers', marker: {size: size + 3, color: 'rgba(187,187,187,0.9)', line:{color: '#222', width: []}}, showlegend:false, name: "unknown"},
    // light blue
    {x:[], y:[], text: [], type: 'scatter', mode: 'markers', marker: {size: size + 3, color: 'rgba(153,204,255,0.9)', line:{color: '#222', width: []}}, showlegend:true, name: "male"},
    // light red
    {x:[], y:[], text: [], type: 'scatter', mode: 'markers', marker: {size: size + 3, color: 'rgba(255,153,170,0.9)', line:{color: '#222', width: []}}, showlegend:true, name: "female"},
]
var xf = jQuery('#plotbx_select').val()
var yf = jQuery('#plotby_select').val()

for (i in sample_data) {
    var idx = 0
    var sex = sample_data[i].sex
    if (sex == "male") {
        idx = 1
    } else if (sex == "female") {
        idx = 2
    }
    sample_data[i].heterozygosity = sample_data[i].n_het / (sample_data[i].n_het + sample_data[i].n_hom_ref + sample_data[i].n_hom_alt)

    traces_b[idx].x.push(sample_data[i][xf])
    traces_b[idx].y.push(sample_data[i][yf])
    traces_b[idx].text.push("sample:" + sample_data[i].sample)

    // // individual is male; highlight points of low Y coverage
    // if (idx == 1) {
    //     // these samples could be female mislabeled as male
    //     if (sample_data[i]["y_depth_mean"] <= 0.2) {
    //         traces_b[idx].marker.line.width.push(size < 7 ? 1.5 : 2)
    //     } else {
    //         traces_b[idx].marker.line.width.push(0)
    //     }
    // // handle unknown and female
    // } else {
    //     // these samples could be male and mislabeled as female
    //     if (sample_data[i]["y_depth_mean"] > 0.2) {
    //         traces_b[idx].marker.line.width.push(size < 7 ? 1.5 : 2)
    //     } else {
    //         traces_b[idx].marker.line.width.push(0)
    //     }
    // }
}

var pa = document.getElementById("plota")
var pb = document.getElementById("plotb")
Plotly.newPlot(pa, input, layout_a)

pa.on('plotly_hover', function(e) {
    var p = e.points[0]
    var split = p.text.split("<br>", 2)

    Plotly.Fx.hover(pb,
        [
            {curveNumber: 0, pointNumber: unknown_sample_names.indexOf(split[0])},
            {curveNumber: 0, pointNumber: unknown_sample_names.indexOf(split[1])},
            {curveNumber: 1, pointNumber: male_sample_names.indexOf(split[0])},
            {curveNumber: 1, pointNumber: male_sample_names.indexOf(split[1])},
            {curveNumber: 2, pointNumber: female_sample_names.indexOf(split[0])},
            {curveNumber: 2, pointNumber: female_sample_names.indexOf(split[1])},
        ]
    )
})

init_depth_metrics_plot()

pa.on('plotly_unhover', function(e) {
    Plotly.Fx.unhover(pb)
})

function pa_hover(sample) {
    var ht = []
    var xs = []
    var ys = []
    if (sample) {
        var left = sample + "<br>"
        var right = "<br>" + sample
        var otr
        for (var j=0; j < input.length-1; j++) {
            otr = input[j]
 
            for (var k=0; k < otr.x.length; k++) {
                if (otr.text[k].startsWith(left) || otr.text[k].endsWith(right)) {
                    xs.push(otr.x[k])
                    ys.push(otr.y[k])
                    ht.push(otr.text[k])
                }
            }
        }
    }

    input[input.length-1].x = xs
    input[input.length-1].y = ys
    input[input.length-1].hovertext = ht
    input[input.length-1].marker.opacity = 1

    // remove all existing markers so highlights will be drawn on top
    //Plotly.purge(pa)
    // redraw with existing + marker traces
    Plotly.react(pa, input, layout_a)
}

function select_and_hover() {
    // get selection
    var sample = sample_search_obj.items[0]
    //console.log("in select and hover with", sample)
    
    if (sample) {
        // get curve and sample index
        for (let curve_idx=0; curve_idx < pb.data.length; curve_idx++) {
            for (let sample_idx=0; sample_idx < pb.data[curve_idx].text.length; sample_idx++) {
                let b_sample = pb.data[curve_idx].text[sample_idx]
                b_sample = b_sample.substring(7, b_sample.length)
                if (b_sample == sample) {
                    Plotly.Fx.hover(pb, [
                        {curveNumber: curve_idx, pointNumber: sample_idx}
                    ])
                    pa_hover(sample)
                }
            }
        }
    } else {
        // sample de-selected; reset A
        pa_hover(false)
        // reset B
        Plotly.purge(pb)
        init_depth_metrics_plot()
    }
}

function init_depth_metrics_plot() {
    Plotly.react(pb, traces_b, layout_b)

    pb.removeAllListeners('plotly_hover')
    pb.removeAllListeners('plotly_unhover')

    pb.on('plotly_hover',  function(e) {
        var p = e.points[0]
        if (p.text[6] != ":") {
            alert("bad sample name:" + p.text[6])
        }
        var sample = p.text.substring(7, p.text.length)
        pa_hover(sample)
    })

    // remove the markers and fix opacity on unhover
    pb.on('plotly_unhover',  function(e) {
        input[input.length-1].x = [];
        input[input.length-1].y = [];
        input[input.length-1].hovertext = [];
        for(j = 0; j < input.length-1; j++){
            input[j].marker.opacity = 1
        }
        Plotly.react(pa, input, layout_a)
    })
}

// select listeners for plot A
jQuery('#plotay_select, #plotax_select, #n_scale').on('change', function() {
    pa_hover(false)
    var scale = jQuery('#n_scale').is(":checked");
    if(this.id != "n_scale") {
        var metric = this.value
        set_xy_data_by_group(input, metric, this.id == "plotax_select", scale)
    } else {
        // if scale, then we need to reset both X and Y
        set_xy_data_by_group(input, jQuery("#plotax_select").val(), true, scale)
        set_xy_data_by_group(input, jQuery("#plotay_select").val(), false, scale)
   }


    layout_a.yaxis.title = jQuery("#plotay_select option:selected").text();
    var yv = jQuery('#plotay_select').val()
    if(scale && yv != "relatedness" && yv != "concordance") { layout_a.yaxis.title = "Scaled " + layout_a.yaxis.title }

    var xv = jQuery('#plotax_select').val()
    layout_a.xaxis.title = jQuery("#plotax_select option:selected").text();
    if(scale && xv != "relatedness" && xv != "concordance") { layout_a.xaxis.title = "Scaled " + layout_a.xaxis.title }
    Plotly.react(pa, input, layout_a)
    select_and_hover()
})

// select listeners for plot B
jQuery('#plotbx_select, #plotby_select').on('change', function() {
    var field = this.value
    var extracted = [[], [], []]
    for (i in sample_data) {
        var sex = sample_data[i].sex
        var idx = sex == "male" ? 1 : ((sex == "female")? 2: 0)
        extracted[idx].push(sample_data[i][field])
    }
    if (this.id == 'plotbx_select') {
        layout_b.xaxis.title = jQuery("#plotbx_select option:selected").text()
        traces_b[0].x = extracted[0]
        traces_b[1].x = extracted[1]
        traces_b[2].x = extracted[2]
    } else {
        layout_b.yaxis.title = jQuery("#plotby_select option:selected").text()
        traces_b[0].y = extracted[0]
        traces_b[1].y = extracted[1]
        traces_b[2].y = extracted[2]
    }
    Plotly.react(pb, traces_b, layout_b)

    // check if sample is selected in sample-select dropdown
    select_and_hover()
})

// responsive plots
window.onresize = function() {
    Plotly.Plots.resize('plota')
    Plotly.Plots.resize('plotb')
};

const sample_names = sample_data.map(function(x) {
    return x.sample
})
const male_sample_names = sample_data.filter(function(x) { return x.sex == "male"}).map(function(x) { return x.sample })
const female_sample_names = sample_data.filter(function(x) { return x.sex == "female"}).map(function(x) { return x.sample })
const unknown_sample_names = sample_data.filter(function(x) { return x.sex != "male" && x.sex != "female"}).map(function(x) { return x.sample })

// sample search
const sample_list = sample_data.map(function(x) {
    return { item: x.sample }
})
const sample_search = $('#sample-search').selectize({
    plugins: ['remove_button'],
    valueField: 'item',
    labelField: 'item',
    searchField: 'item',
    options: sample_list,
    placeholder: 'Sample ID',
    mode: 'single',
    closeAfterSelect: false,
})
const sample_search_obj = sample_search[0].selectize

$('#sample-search').on('change', function() {
    select_and_hover()
})
</script>

</body>
</html>
