{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from cyvcf2 import VCF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "dir_processed_sample_vcfs = '/scratch/ucgd/lustre-work/quinlan/data-shared/datasets/spermseq/final_seq/19610R_batch1and2_combined_wf/quinlan_analysis/custom/processed_vcfs/sample_vcfs'\n",
    "fp_metadata = '/scratch/ucgd/lustre-work/quinlan/data-shared/datasets/spermseq/final_seq/19610R_batch1and2_combined_wf/quinlan_analysis/custom/metadata/reliable_samples_MRs.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>donor_ID</th>\n",
       "      <th>sample_ID</th>\n",
       "      <th>input_DNA</th>\n",
       "      <th>reliable</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>analysis</th>\n",
       "      <th>type</th>\n",
       "      <th>age</th>\n",
       "      <th>tissue_type</th>\n",
       "      <th>mean_no_call_rate</th>\n",
       "      <th>...</th>\n",
       "      <th>genomic_mutation_rate_SNV_intronic</th>\n",
       "      <th>genomic_mutation_rate_SNV_exonic</th>\n",
       "      <th>genomic_mutation_rate_INDEL_intronic</th>\n",
       "      <th>genomic_mutation_rate_INDEL_exonic</th>\n",
       "      <th>num_mutations</th>\n",
       "      <th>num_unique_inherited_mutations</th>\n",
       "      <th>num_clonal_mutations</th>\n",
       "      <th>num_non_clonal_mutations</th>\n",
       "      <th>num_pathogenic_mutations</th>\n",
       "      <th>duplex_bases</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1SCS014</td>\n",
       "      <td>19610X1.1</td>\n",
       "      <td>200 ng</td>\n",
       "      <td>True</td>\n",
       "      <td>4/23/10</td>\n",
       "      <td>cross-sectional</td>\n",
       "      <td>paired samples</td>\n",
       "      <td>35.9</td>\n",
       "      <td>Sperm DNA</td>\n",
       "      <td>0.118964</td>\n",
       "      <td>...</td>\n",
       "      <td>2.153297e-08</td>\n",
       "      <td>3.866146e-08</td>\n",
       "      <td>9.787712e-10</td>\n",
       "      <td>9.787712e-10</td>\n",
       "      <td>127</td>\n",
       "      <td>13</td>\n",
       "      <td>2</td>\n",
       "      <td>125</td>\n",
       "      <td>0</td>\n",
       "      <td>2043378475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1SCS014</td>\n",
       "      <td>19610X2.1</td>\n",
       "      <td>200 ng</td>\n",
       "      <td>True</td>\n",
       "      <td>4/23/10</td>\n",
       "      <td>cross-sectional</td>\n",
       "      <td>paired samples</td>\n",
       "      <td>35.9</td>\n",
       "      <td>Blood DNA</td>\n",
       "      <td>0.100579</td>\n",
       "      <td>...</td>\n",
       "      <td>3.828919e-08</td>\n",
       "      <td>7.864807e-08</td>\n",
       "      <td>3.104529e-09</td>\n",
       "      <td>2.069686e-09</td>\n",
       "      <td>236</td>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "      <td>233</td>\n",
       "      <td>7</td>\n",
       "      <td>1932660251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>D091_IX8Y012</td>\n",
       "      <td>19610X13.1</td>\n",
       "      <td>200 ng</td>\n",
       "      <td>True</td>\n",
       "      <td>3/24/95</td>\n",
       "      <td>longitudinal</td>\n",
       "      <td>1st collection</td>\n",
       "      <td>52.0</td>\n",
       "      <td>Sperm DNA</td>\n",
       "      <td>0.092290</td>\n",
       "      <td>...</td>\n",
       "      <td>2.013937e-08</td>\n",
       "      <td>3.812095e-08</td>\n",
       "      <td>7.192633e-10</td>\n",
       "      <td>4.315580e-09</td>\n",
       "      <td>88</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>88</td>\n",
       "      <td>3</td>\n",
       "      <td>1390311543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>D091_IX8Y012</td>\n",
       "      <td>19610X14.1</td>\n",
       "      <td>200 ng</td>\n",
       "      <td>True</td>\n",
       "      <td>9/6/11</td>\n",
       "      <td>longitudinal</td>\n",
       "      <td>last collection</td>\n",
       "      <td>68.0</td>\n",
       "      <td>Blood DNA</td>\n",
       "      <td>0.065349</td>\n",
       "      <td>...</td>\n",
       "      <td>1.390663e-07</td>\n",
       "      <td>2.074406e-07</td>\n",
       "      <td>2.317772e-09</td>\n",
       "      <td>2.317772e-09</td>\n",
       "      <td>303</td>\n",
       "      <td>2</td>\n",
       "      <td>12</td>\n",
       "      <td>291</td>\n",
       "      <td>1</td>\n",
       "      <td>862897462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>D091_IX8Y012</td>\n",
       "      <td>19610X15.1</td>\n",
       "      <td>200 ng</td>\n",
       "      <td>True</td>\n",
       "      <td>9/20/11</td>\n",
       "      <td>longitudinal</td>\n",
       "      <td>last collection</td>\n",
       "      <td>68.0</td>\n",
       "      <td>Sperm DNA</td>\n",
       "      <td>0.086492</td>\n",
       "      <td>...</td>\n",
       "      <td>2.149925e-08</td>\n",
       "      <td>5.486015e-08</td>\n",
       "      <td>2.965414e-09</td>\n",
       "      <td>4.448120e-09</td>\n",
       "      <td>113</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>113</td>\n",
       "      <td>5</td>\n",
       "      <td>1348884351</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        donor_ID   sample_ID input_DNA  reliable collection_date  \\\n",
       "0        1SCS014   19610X1.1    200 ng      True         4/23/10   \n",
       "1        1SCS014   19610X2.1    200 ng      True         4/23/10   \n",
       "10  D091_IX8Y012  19610X13.1    200 ng      True         3/24/95   \n",
       "11  D091_IX8Y012  19610X14.1    200 ng      True          9/6/11   \n",
       "12  D091_IX8Y012  19610X15.1    200 ng      True         9/20/11   \n",
       "\n",
       "           analysis             type   age tissue_type  mean_no_call_rate  \\\n",
       "0   cross-sectional   paired samples  35.9   Sperm DNA           0.118964   \n",
       "1   cross-sectional   paired samples  35.9   Blood DNA           0.100579   \n",
       "10     longitudinal   1st collection  52.0   Sperm DNA           0.092290   \n",
       "11     longitudinal  last collection  68.0   Blood DNA           0.065349   \n",
       "12     longitudinal  last collection  68.0   Sperm DNA           0.086492   \n",
       "\n",
       "    ...  genomic_mutation_rate_SNV_intronic  genomic_mutation_rate_SNV_exonic  \\\n",
       "0   ...                        2.153297e-08                      3.866146e-08   \n",
       "1   ...                        3.828919e-08                      7.864807e-08   \n",
       "10  ...                        2.013937e-08                      3.812095e-08   \n",
       "11  ...                        1.390663e-07                      2.074406e-07   \n",
       "12  ...                        2.149925e-08                      5.486015e-08   \n",
       "\n",
       "    genomic_mutation_rate_INDEL_intronic  genomic_mutation_rate_INDEL_exonic  \\\n",
       "0                           9.787712e-10                        9.787712e-10   \n",
       "1                           3.104529e-09                        2.069686e-09   \n",
       "10                          7.192633e-10                        4.315580e-09   \n",
       "11                          2.317772e-09                        2.317772e-09   \n",
       "12                          2.965414e-09                        4.448120e-09   \n",
       "\n",
       "    num_mutations  num_unique_inherited_mutations  num_clonal_mutations  \\\n",
       "0             127                              13                     2   \n",
       "1             236                              13                     3   \n",
       "10             88                               2                     0   \n",
       "11            303                               2                    12   \n",
       "12            113                               2                     0   \n",
       "\n",
       "    num_non_clonal_mutations  num_pathogenic_mutations  duplex_bases  \n",
       "0                        125                         0    2043378475  \n",
       "1                        233                         7    1932660251  \n",
       "10                        88                         3    1390311543  \n",
       "11                       291                         1     862897462  \n",
       "12                       113                         5    1348884351  \n",
       "\n",
       "[5 rows x 27 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samps_df = pd.read_csv(fp_metadata)\n",
    "reliable_samps_df = samps_df.loc[samps_df['reliable']]\n",
    "reliable_samps_df[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "### subset metadata df to only sperm samples\n",
    "reliable_sperm_samps_df = reliable_samps_df.loc[reliable_samps_df['tissue_type'] == 'Sperm DNA']\n",
    "\n",
    "### subset metadata df to only blood samples\n",
    "reliable_blood_samps_df = reliable_samps_df.loc[reliable_samps_df['tissue_type'] == 'Blood DNA']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr4 54732043 54732042 54732044 TG ['T']\n",
      "indel\n",
      "shared_blood_sperm;num_donors_with_mut_gt_1;\n",
      "alt_depth:  1\n",
      "VAF:  0.0007824726\n",
      "context:  TTG\n"
     ]
    }
   ],
   "source": [
    "vcf = list(VCF('/scratch/ucgd/lustre-work/quinlan/data-shared/datasets/spermseq/final_seq/19610R_batch1and2_combined_wf/quinlan_analysis/custom/processed_vcfs/sample_vcfs/19610X1.1/19610X1.1.variants_filtered.vcf.gz'))\n",
    "variant = vcf[123]\n",
    "# *** variant.POS gives you vcf coordinate; variant.start and variant.end give you bed coordinates)\n",
    "print(variant.CHROM, variant.POS, variant.start, variant.end, variant.REF, variant.ALT)\n",
    "print(variant.INFO.get('variation_type'))\n",
    "print(variant.format('filter')[0]) # this will either be PASS or a semi-colon separated list of filters this variant failed\n",
    "print('alt_depth: ', variant.format('alt_depth')[0][0])\n",
    "print('VAF: ', variant.format('AF')[0][0])\n",
    "print('context: ', variant.format('context')[0])\n",
    "\n",
    "# *** variant.format('some_field') returns a list of values, one for each sample represented in the VCF. since the sample vcfs only have one sample column, always use index [0] for those. If using the donor vcfs or any vcf with multiple sample columns index accordingly\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1 243572952 C>T CCG\n",
      "chr2 47798625 C>T ACG\n",
      "chr3 9756552 C>T ACG\n",
      "chr3 41233763 C>T CCG\n",
      "chr5 80873118 G>A CGC\n"
     ]
    }
   ],
   "source": [
    "def is_CpG_to_TpG(v):\n",
    "    return (v.format('context')[0][1:] == 'CG' and v.INFO.get('subtype') == 'C>T') or (v.format('context')[0][1:] == 'GC' and v.INFO.get('subtype') == 'G>A')\n",
    "\n",
    "CpGs = [variant for variant in vcf if is_CpG_to_TpG(variant)]\n",
    "\n",
    "for variant in CpGs[:5]:\n",
    "    print(variant.CHROM, variant.POS, variant.INFO.get('subtype'), variant.format('context')[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "### iterate through reliable samples, get processed VCF, and subset VCF\n",
    "\n",
    "for idx,row in reliable_samps_df.iterrows():\n",
    "\n",
    "    sample_ID = row['sample_ID']\n",
    "    tissue_type = row['tissue_type']\n",
    "\n",
    "    fp_vcf = '{}/{}/{}.variants_filtered.vcf.gz'.format(dir_processed_sample_vcfs, sample_ID, sample_ID) # the '.variants_filtered.vcf.gz' files are easier to look at, but if you want VEP and ClinVar annotations choose 'variants_filtered_annotated.vcf.gz'\n",
    "    sample_vcf = list(VCF(fp_vcf)) # ***if you want to iterate over the VCF object more than once, cast it as a list***\n",
    "\n",
    "    ### get SNVs only\n",
    "    SNVs = [v for v in sample_vcf if v.INFO.get('variation_type') == 'snv']\n",
    "\n",
    "    ### get clonal variants only\n",
    "    clonal = [v for v in sample_vcf if v.format('alt_depth')[0] > 1]\n",
    "\n",
    "    ### get nonclonal variants only\n",
    "    nonclonal = [v for v in sample_vcf if v.format('alt_depth')[0] == 1]\n",
    "\n",
    "    ### get PASS mutations (SNVs and INDELs that pass quality filters and do not recur across samples)\n",
    "    pass_mutations = [v for v in sample_vcf if v.format('filter')[0] == 'PASS']\n",
    "\n",
    "    ### get PASS mutations AND mutations that recur across samples (can be recurrent in any tissue)\n",
    "    allow_shared_donor = [v for v in sample_vcf if v.format('filter')[0] in ['PASS', 'num_donors_with_mut_gt_1;']]\n",
    "\n",
    "    ### get PASS mutations AND mutations that recur only in samples of the same tissue type (e.g. if sample_ID is a sperm sample, only allow variants that recur in sperm)\n",
    "    other_tissue = 'blood' if tissue_type == 'Sperm DNA'else 'sperm'\n",
    "    allow_shared_tissue = [v for v in sample_vcf if v.format('filter')[0] in ['PASS', 'num_donors_with_mut_gt_1;'] and v.INFO.get('num_donors_with_variant_at_this_pos_in_'+other_tissue)==0]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7.12 ('ss': conda)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "2ffe01c38ece59cc9c45ce27132c0f5ba0d86e879dd1a66975f5e794ff2c7be7"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
