# /scratch/ucgd/lustre-work/quinlan/data-shared/datasets/spermseq/wgs

configfile: '../data/config.yaml'
in_dir: str = config["input_directory"]
out_dir: str = config["output_directory"]
reference: str = config["reference_genome"]
samples: list[str] = config["samples"]



### IMPORTS ###
import snakemake
import os
import re

# Pipeline plans
# - Variant calling (Call mutations)
#     - FreeBayes?
#     - DeepVariant?
#     - Likely SNVs, maybe indels?
# - Pipeline to process called variants/Vcf
#     - PASS variants, filter
# - Mutational signatures
#     - https://github.com/HLee-Six/colon_microbiopsies what Lee-Six did (in R)
# - Mutational spectra
# - Make plots and save to rational places


samples: list[str] = [
    "test"
]

# Wildcard constraints; should all be alphanumeric
wildcard_constraints:
    sample = "[A-Za-z0-9]+"

# Rule 'all', defining all requested output.
rule all:
    input:
        expand("{out_dir}/data/{sample}-sorted.bai", out_dir = out_dir, sample = samples)

# Rule 'fastp' preprocesses the reads by detecting and removing adapters,
# filtering reads with an average quality less than 20, and producing
# filtered fastq reads with preprocessing reports.
rule fastp:
    input:
        R1= in_dir + "/{sample}_R1.fastq.gz",
        R3= in_dir + "/{sample}_R3.fastq.gz"
    output:
        r1_clean = "{out_dir}/{sample}_R1.clean.fastq",
        r3_clean = "{out_dir}/{sample}_R3.clean.fastq",
        html_report = "{out_dir}/reports/{sample}-fastp-report.html",
        json_report = "{out_dir}/reports/{sample}-fastp-report.json"
    shell:
        """
        fastp --in1 {input.R1} --in2 {input.R3} \
        --out1 {output.r1_clean} --out2 {output.r3_clean} \
        --detect_adapter_for_pe \
        --average_qual 20 \
        --thread {threads} \
        --html {output.html_report} \
        --json {output.json_report}
        """

# Aligns cleaned fastq reads to the defined reference genome
# Mark duplicates and extract discordant and split reads from sam files
# Convert to bam (exclude unmapped reads with -F 4)
rule bwa_mem_bam:
    input:
        r1_clean = "{out_dir}/{sample}_R1.clean.fastq",
        r3_clean = "{out_dir}/{sample}_R3.clean.fastq",
        ref = reference
    output:
       zip_bam = "{out_dir}/{sample}.bam.gz"
    threads: 8
    conda: in_dir + "/semicolonsmk.yaml"
    shell:
        "bwa mem -t {threads} {input.ref} {input.r1_clean} {input.r3_clean} | samblaster | samtools view -b -F 4 > {output.zip_bam}"

rule gunzip:
    input:
       zip_bam = "{out_dir}/{sample}.bam.gz"
    output:
        bam = "{out_dir}/{sample}.bam"
    shell:
        """
        gunzip {input.zip_bam} > {output.bam}
        """

rule sort_bam:
    input:
        bam = "{out_dir}/{sample}.bam"
    output:
        bam_sort = "{out_dir}/{sample}-sorted.bam"
    shell:
        """
        samtools sort {input.bam} -o {output.bam_sort}
        """

rule index_bam:
    input:
        bam_sort = "{out_dir}/{sample}-sorted.bam"
    output:
        bai = "{out_dir}/{sample}-sorted.bai"
    shell:
        """
        samtools index -b {input.bam_sort} -o {output.bai}
        """